pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+	// Network is the path to the network namespace.
+	Network string
+}
+
+// LinuxSandBoxStatus contains status specific to Linux sandboxes.
+type LinuxPodSandboxStatus struct {
+	// Namespaces contains paths to the sandbox's namespaces.
+	Namespaces *Namespaces
+}
+
+// PodSandboxResources contains the CPU/memory resource requirements.
+type PodSandboxResources struct {
+	// CPU resource requirement.
+	CPU api.ResourceRequirements
+	// Memory resource requirement.
+	Memory api.ResourceRequirements
+}
+
+// This is to distinguish with existing ContainerID type, which includes a
+// runtime type prefix (e.g., docker://). We may rename this later.
+type RawContainerID string
+
+// ContainerRuntime provides methods for container lifecycle operations, as
+// well as listing or inspecting existing containers.
+type ContainerRuntime interface {
+	// Create creates a container in the sandbox, and returns the ID
+	// of the created container.
+	Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxID PodSandboxID) (RawContainerID, error)
+	// Start starts a created container.
+	Start(RawContainerID ContainerID) error
+	// Stop stops a running container with a grace period (i.e., timeout).
+	Stop(RawContainerID string, timeout int) error
+	// Remove removes the container.
+	Remove(RawContainerID string) error
+	// List lists the existing containers that match the ContainerFilter.
+	// The returned list should only include containers previously created
+	// by this ContainerManager.
+	List(filter ContainerFilter) ([]Container, error)
+	// Status returns the status of the container.
+	Status(RawContainerID string) (ContainerStatus, error)
+}
+
+// ContainerCommandExecutor provides methods to run a command in the container.
+// TODO: Should we merge this with ContainerRuntime?
+type ContainerCommandExecutor interface {
+	// Exec executes a command in the container.
+	Exec(RawContainerID string, cmd []string, streamOpts StreamOptions) error
+}
+
+type ContainerConfig struct {
+	// Name of the container.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// Command to execute (i.e., entrypoint for docker)
+	Command []string
+	// Args for the Command (i.e., command for docker)
+	Args []string
+	// Current working directory of the command.
+	WorkingDir string
+	// List of environment variable to set in the container
+	Env []KeyValue
+	// Mounts specifies mounts for the container
+	Mounts []Mount
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Privileged runs the container in the privileged mode.
+	Privileged bool
+	// ReadOnlyRootFS sets the root filesystem of the container to be
+	// read-only.
+	ReadOnlyRootFS bool
+	// Path to store the container log on the host (i.e., outside of the
+	// sandbox).
+	LogPath string
```

Comment:
> My concerns are echoed in that issue as well; a file-path doesn't play well with journald (nor other logging solutions IMO).

Why does `journald` have to be in the critical path between a container and its logs? Users of kuberentes will most likely run a cluster level logging service and/or use the `kubectl logs` interface to consume logs. As an end user of kubernetes, why would one care about journald? Journald does not seem to provide the isolation that is required for kubernetes. 

> Will that actually occur in practice? Right now, both Docker and rkt store logs in locations of their own choosing in their own formats (json and journald respectively). If you require the runtimes to store logs in one additional location in a plainer format (presumably logs.stderr, logs.stdout), you'll end up with two copies of the logs, only one of which is managed as you want. Perhaps I'm misunderstanding something there.

This proposal is not meant to just provide compatibility with all existing FOSS. It is rather meant to define how a runtime for kubernetes should be. Continuing to design against all the various design assumptions being made in various FOSS projects is not really adding much value to k8s end users.
In the near future, we might have new runtimes that meet all the requirements of this API and provide better integration with Kubernetes. 
As for the existing runtimes, we will document known issues with each runtime and let users make their choice. So if docker cannot provide better isolation for log files, it is up to the user to decide if they want to use docker as the runtime.
As for docker support, kubelet creates symlinks to the log files being created by docker. We can continue doing that, even though it does not meet all the requirements.

> Some users may want to log directly to a remote server. Some users may want to take advantage of logging features their container runtimes provide (e.g. additional metadata). So long as the runtime can provide a streaming interface to stdout/err, why should we get in the way of doing such things?

The main reason is to provide a stable extension point for logging agents to collect logs from kubernetes pods. By letting each runtime handle logs in its own manner, we will have a pretty fragmented ecosystem, with logging vendors not knowing how (& where) to integrate with k8s.
A standalone use case of a container runtime might not match completely with that of k8s use case.